pkg/model/multiversion/delta.go

```diff
@@ -0,0 +1,248 @@
+				(hubField.FieldConfig != nil && hubField.FieldConfig.IsSecret) {
+				deltas = append(deltas, FieldDelta{
+					Spoke:      spokeField,
+					Hub:        hubField,
+					ChangeType: FieldChangeTypeShapeChangedToSecret,
+				})
+				continue
+			}
+
+			// TODO(a-hilaly) better function for equalizing shapes
+			equalShapes, err := isEqualShape(spokeField.ShapeRef, hubField.ShapeRef)
+			if err != nil {
+				return nil, err
+			}
+			if equalShapes {
+				// if the fields have equal names and types the change is intact
+				deltas = append(deltas, FieldDelta{
+					Spoke:      spokeField,
+					Hub:        hubField,
+					ChangeType: FieldChangeTypeIntact,
+				})
+				continue
+			}
+
+			// at this point we know that the fields kept the same name but have different
+			// shapes
+			deltas = append(deltas, FieldDelta{
+				Spoke:      spokeField,
+				Hub:        hubField,
+				ChangeType: FieldChangeTypeIntact,
+			})
+			continue
+		}
+
+		// if a field is not found in the hubFieldsNames, there are three
+		// possible changes: Removed, Added or Renamed.
+
+		// First let's check if field was renamed
+		newName, ok := renames[spokeFieldName]
+		if ok {
+			hubField, ok2 := hubCRDFields[newName]
+			if !ok2 {
+				// if a field was renamed and we can't find it in hubFieldsNames, something
+				// very wrong happend during CRD loading.
+				return nil, fmt.Errorf("cannot find renamed field %s " + newName)
+			}
+
+			// mark field as visited, both old and new names.
+			visitedFields[newName] = nil
+			visitedFields[spokeFieldName] = nil
+
+			// this will mostlikely be always true, but let's double check.
+			if newName == hubField.Names.Camel {
+				// field was renamed
+				deltas = append(deltas, FieldDelta{
+					Spoke:      spokeField,
+					Hub:        hubField,
+					ChangeType: FieldChangeTypeRenamed,
+				})
+				continue
+			}
+			return nil, fmt.Errorf("renamed field unmatching: %v != %v", newName, hubField.Names.Camel)
+		}
+
+		// If the field was not renamed nor left intact nor it shape changed, it's
+		// a removed field.
+		deltas = append(deltas, FieldDelta{
+			Spoke:      spokeField,
+			Hub:        nil,
+			ChangeType: FieldChangeTypeRemoved,
+		})
+	}
+
+	// At this point we collected every type of change except added fields.
+	// To find added fields we just look for fields that are in hubFieldsNames
+	// and were not visited before (are not in spokeFieldsNames).
+	for _, hubFieldName := range hubFieldsNames {
+		_, visited := visitedFields[hubFieldName]
+		if visited {
+			continue
+		}
+
+		hubField, _ := hubCRDFields[hubFieldName]
+		deltas = append(deltas, FieldDelta{
+			Spoke:      nil,
+			Hub:        hubField,
+			ChangeType: FieldChangeTypeAdded,
+		})
+	}
+
+	return deltas, nil
+}
+
+// isEqualShape returns whether two awssdkmodel.ShapeRef are equal or not.
+// TODO(a-hilaly): this is very fragile - a simple docstring change will make the
+// result wrong - we'll need to slowly verify each member/key/value
```

Comment:
Can we pass this through our existing generator methods to convert it to Go types before comparing?

Reply:
@RedbackThomson  @jaypipes `cleanGoType` will always return the same type name even if the shape changes. I think that we need to minutely compare the shape members and their types.

Reply:
Yeah, I see what you're saying @A-Hilaly. ++